File: backend/app/schemas/world.py
"""Schemas for the world map: features, players, custom markers."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

FeatureType = Literal[
    "factory",
    "resource_node",
    "resource_well",
    "geyser",
    "power_plant",
    "train_station",
    "drone_port",
    "truck_station",
    "artifact",
    "collectible",
    "wreck",
]


class Position(BaseModel):
    """Game-world position in centimeters (Unreal units); +x east, +y south."""

    x: float
    y: float
    z: float = 0


class MapFeature(BaseModel):
    """A world object shown on the map."""

    id: str
    type: FeatureType
    name: str
    position: Position
    meta: dict[str, str | float | int] = Field(default_factory=dict)
    collected: bool | None = Field(
        default=None, description="Pickups only: already collected by a player"
    )
    occupied: bool | None = Field(
        default=None, description="Resource nodes only: an extractor is installed"
    )


class PlayerInfo(BaseModel):
    """A player and their live position."""

    id: str
    name: str
    position: Position
    online: bool = True


class WorldSnapshot(BaseModel):
    """Static features plus live player positions."""

    generated_at: datetime
    source: Literal["simulation", "frm", "save"]
    players: list[PlayerInfo]
    features: list[MapFeature]


class CustomMarkerIn(BaseModel):
    """User-created marker payload."""

    name: str = Field(min_length=1, max_length=128)
    icon: str = Field(default="pin", max_length=32)
    color: str = Field(default="#9085e9", pattern=r"^#[0-9a-fA-F]{6}$")
    position: Position
    notes: str = Field(default="", max_length=2000)


class CustomMarker(CustomMarkerIn):
    """A persisted custom marker."""

    id: int

File: frontend/src/types/world.ts
/** Mirrors of backend/app/schemas/world.py. */

export type FeatureType =
  | 'factory'
  | 'resource_node'
  | 'resource_well'
  | 'geyser'
  | 'power_plant'
  | 'train_station'
  | 'drone_port'
  | 'truck_station'
  | 'artifact'
  | 'collectible'
  | 'wreck';

export interface Position {
  x: number;
  y: number;
  z: number;
}

export interface MapFeature {
  id: string;
  type: FeatureType;
  name: string;
  position: Position;
  meta: Record<string, string | number>;
  /** Pickups only: already collected by a player. */
  collected: boolean | null;
  /** Resource nodes only: an extractor is installed. */
  occupied: boolean | null;
}

export interface PlayerInfo {
  id: string;
  name: string;
  position: Position;
  online: boolean;
}

/** A conveyor belt segment rendered as a polyline on the map. */
export interface BeltPath {
  id: string;
  name: string;
  /** Short UE class, e.g. Build_ConveyorBeltMk1_C. */
  class_name: string;
  /** Spline points in cm. */
  points: Position[];
  items_per_minute: number | null;
}

export interface WorldSnapshot {
  generated_at: string;
  source: 'simulation' | 'frm' | 'save';
  players: PlayerInfo[];
  features: MapFeature[];
  belts: BeltPath[];
  /** Power lines (straight segments). */
  cables: BeltPath[];
  /** Pipeline splines. */
  pipes: BeltPath[];
}

export interface CustomMarkerIn {
  name: string;
  icon?: string;
  color?: string;
  position: Position;
  notes?: string;
}

export interface CustomMarker extends Required<CustomMarkerIn> {
  id: number;
}
